src/utill.rs

```diff
@@ -637,17 +645,69 @@ pub(crate) fn check_tor_status(control_port: u16, password: &str) -> Result<(),
     Ok(())
 }
 
-pub(crate) fn get_tor_hostname() -> Result<String, TorError> {
-    let path = if cfg!(target_os = "macos") {
-        "/opt/homebrew/var/lib/tor/coinswap/hostname"
-    } else {
-        "/var/lib/tor/coinswap/hostname"
-    };
+pub(crate) fn get_emphemeral_address(
+    control_port: u16,
+    target_port: u16,
+    password: &str,
+) -> Result<String, TorError> {
+    use std::io::BufRead;
+    let mut stream = TcpStream::connect(format!("127.0.0.1:{}", control_port))?;
+    let mut reader = BufReader::new(stream.try_clone()?);
+    let auth_command = format!("AUTHENTICATE \"{}\"\r\n", password);
+    stream.write_all(auth_command.as_bytes())?;
+    let add_onion_command = format!(
+        "ADD_ONION NEW:BEST Port={},127.0.0.1:{}\r\n",
+        target_port, target_port
+    );
+    stream.write_all(add_onion_command.as_bytes())?;
+    let mut response = String::new();
+    let mut service_id = String::new();
+
+    while reader.read_line(&mut response)? > 0 {
+        if response.starts_with("250-ServiceID=") {
+            service_id = response
+                .trim_start_matches("250-ServiceID=")
+                .trim()
+                .to_string();
+            break;
+        }
+        response.clear();
+    }
+
+    if service_id.is_empty() {
+        return Err(TorError::General(
+            "Failed to retrieve ephemeral onion service details".to_string(),
+        ));
+    }
+    Ok(format!("{}.onion", service_id))
+}
+
+pub(crate) fn get_tor_hostname(
+    datadir: PathBuf,
+    control_port: u16,
+    target_port: u16,
+    password: &str,
+) -> Result<String, TorError> {
+    let tor_config_path = datadir.join("tor/hostname");
+
+    if tor_config_path.exists() {
+        if let Ok(hostname) = fs::read(&tor_config_path) {
+            let hostname: String = serde_cbor::de::from_slice(&hostname)?;
+            let hostname = hostname.trim().to_string();
+            log::info!("Loaded existing Tor hostname: {}", hostname);
+            return Ok(hostname);
+        }
+    }
+
+    let hostname = get_emphemeral_address(control_port, target_port, password)?;
```

Comment:
Why is this thing ephemeral? We will be needing a persistent onion service? 

Seems like this will generate a new hostname everytime it is called?

Reply:
This will only run when the hostname file doesn't exist. So If we stop the server and start again, the hostname exists, but onion service will not start.